### PR TITLE
Spilling to disk for aggregation v2

### PR DESCRIPTION
**Pull request by pnowojski**
Supersedes https://github.com/prestodb/presto/pull/5142. Notable changes:
- reordered commits
- removed `max-entries-before-spill` property

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/spiller/Spiller.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+package com.facebook.presto.spi.spiller;
+
+import com.facebook.presto.spi.Page;
+
+import java.io.Closeable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+public interface Spiller
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/spiller/Spiller.java]: move this into presto-main as it is no longer used outside

File: presto-main/src/main/java/com/facebook/presto/operator/spiller/BinarySpillerFactory.java [hunk 2]
```diff
@@ -33,28 +33,27 @@
...
-        SPILL_PATH.toFile().mkdirs();
+        this.spillPath = featuresConfig.getSpillerSpillPath();
+        this.spillPath.toFile().mkdirs();
     }
 
     @Override
     public Spiller create(List<Type> types)
     {
-        return new BinaryFileSpiller(blockEncodingSerde, executor);
+        return new BinaryFileSpiller(blockEncodingSerde, executor, spillPath);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/spiller/BinarySpillerFactory.java]: I think each BinaryFileSpiler should get an unique spill path. In case where are two spilling aggregation running concurrently then they might override spills to each other.

File: presto-spi/src/main/java/com/facebook/presto/spi/spiller/Spiller.java (hunk 1 above)
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/spiller/Spiller.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/spiller/BinarySpillerFactory.java (hunk 2 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/spiller/BinarySpillerFactory.java]: yes, each of them gets unique directory. `BinaryFileSpiller` instance is one per operator.

**Comment by pnowojski** [comment]
@cberner just in case. You can divide reviewing and merging process of this PR on this commits:

- Add `experimental.spiller-threads` property (#267)
- Add support for merge with memory for aggregation spill
- Add spilled data size stat to BinaryFileSpiller

**Comment by cberner** [comment]
Cool, thanks! Will take a look at this once I finish the Accumulo review, as that one's been waiting even longer

**Comment by cberner** [comment]
Checked with @haozhun and he's reviewing https://github.com/prestodb/presto/pull/5825, so I'll skip the commits in that PR and start on the ones after

**Comment by pnowojski** [comment]
I have rebased it on the latest master (with merge code from https://github.com/prestodb/presto/pull/5825), so all commits from this PR now belongs to this PR :)

File: presto-main/src/main/java/com/facebook/presto/spiller/SpillerFactory.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+package com.facebook.presto.spiller;
+
+import com.facebook.presto.spi.type.Type;
+
+import java.util.List;
+
+public interface SpillerFactory
+{
+    com.facebook.presto.spiller.Spiller create(List<Type> types);
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/SpillerFactory.java]: nope, some Intellij artifact

File: presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+public interface Spiller
+        extends Closeable
+{
+    /**
+     * Initiate spilling of pages stream. Returns completed future once spilling has finished.
+     */
+    CompletableFuture<?> spill(Iterator<Page> pageIterator);
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java]: Let's not overgeneralize it at this point. We can refactor it later, when such thing will be indeed required.

For joins it seems like we need different implementation of spillers. You can look at our WiP interfaces for join here:
https://github.com/Teradata/presto/pull/356/commits/db6a9b7ca9fa231a7c923fc7af20772c61366afb
and here:
https://github.com/Teradata/presto/pull/356/commits/97c3ebe1baa4f03a6139c877268a9a4e1c687820

We are thinking about later reimplementing this spiller (used in aggregations) so that it will use `SingleStreamSpiller` interface (similarly as `GenericPartitioningSpiller` can wrap any given `SingleStreamSpiller` to provide `PartitioningSpiller`).

I don't see an advantage of having `Slice` as an interface. It would only complicate things for both joins and aggregations.

File: presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Returns list of previously spilled Pages streams.
+     */
+    List<Iterator<Page>> getSpills();
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java]: This shouldn't be that big an issue as for writing, because in writing we have huge chunks of data that we want to write at once. For reading, we are reading Page by Page from different Iterators one by one - one call to `HashAggregationOperator::getOutput()` should request at most for only couple of Page reads from disks. Read ahead should mitigate this issue even further.

What's more is that implementing this in async fashion is not that simple. It would have to be `List<Iterator<Future<Page>>>` and whole `MergeSorter` would have to handle those Futures. Maybe that would force us to write our own `MergeSort`.

I'm not saying this shouldn't be done eventually, but it doesn't seem as the highest priority at the moment. Similarly as Presto was able to live for very long time with synchronous `PageSinks` and `PageSources` (btw, don't that vast majority of connectors, if not even all of the open source ones, still use synchronous `PageSinks/Sources`?).

File: presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+
+public class BinarySpillerFactory
+        implements SpillerFactory
+{
+    public static final Path SPILL_PATH = Paths.get("/tmp/spills");
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java]: It is in later commit.

File: presto-main/src/test/java/com/facebook/presto/spiller/BinaryFileSpillerTest.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.lang.Double.doubleToLongBits;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static org.testng.Assert.assertEquals;
+
+public class BinaryFileSpillerTest
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/spiller/BinaryFileSpillerTest.java]: ops, fixed

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            spiller.get().close();
+        }
+    }
+
+    private CompletableFuture<?> spillToDisk()
+    {
+        hashAggregationBuilder.setOutputPartial();
+
+        if (!spiller.isPresent()) {
+            spiller = Optional.of(spillerFactory.get().create(hashAggregationBuilder.buildTypes()));
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 9]
```diff
@@ -1944,7 +1947,9 @@ private PhysicalOperation planGroupByAggregation(AggregationNode node, PhysicalO
                     hashChannel,
                     node.getGroupIdSymbol().map(mappings::get),
                     10_000,
-                    maxPartialAggregationMemorySize);
+                    maxPartialAggregationMemorySize,
+                    memoryLimitBeforeSpill,
+                    Optional.of(spillerFactory));
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: `Optional.empty()` was used in tests that didn't use spilling. I have replaced that with some inlined implementation that `throw new UnsupportedOperatorException()`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 10]
```diff
@@ -1933,6 +1933,12 @@ private PhysicalOperation planGroupByAggregation(AggregationNode node, PhysicalO
                     .map(entry -> source.getTypes().get(entry))
                     .collect(toImmutableList());
 
+            if (memoryLimitBeforeSpill.toBytes() > 0) {
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: In runtime check queries would nondeterministicaly fail with unsupported exception, and they would fail almost always with current version (with small memory limits before spill, prior to revocable memory). 

Since it's currently an experimental feature, that we expect to be turned off by default, I would prefer to fail fast and in deterministic fashion. At least until revocable memory is added.

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSpiller.java [hunk 11]
```diff
@@ -51,14 +52,20 @@
...
 
     private int spillsCount;
     private final ListeningExecutorService executor;
 
-    public BinaryFileSpiller(BlockEncodingSerde blockEncodingSerde, ListeningExecutorService executor, Path spillPath)
+    public BinaryFileSpiller(
+            BlockEncodingSerde blockEncodingSerde,
+            ListeningExecutorService executor,
+            Path spillPath,
+            AtomicLong spilledDataSize)
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSpiller.java]: Can you elaborate? What's your idea? Reversing the interface would require that Factory keeps track of all created spillers or I would have to add some new proxy interface that would basically be an AtomicLong wrapper.

I was thinking about introducing `SpillingStats` object, but as long it is single long it seemed like an overkill.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 12]
```diff
@@ -393,7 +394,10 @@ protected void setup(Binder binder)
         binder.bind(FinalizerService.class).in(Scopes.SINGLETON);
 
         // Spiller
-        binder.bind(SpillerFactory.class).to(BinarySpillerFactory.class).in(Scopes.SINGLETON);
+        install(installModuleIf(
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: Initially this was implemented as a plugin, but during our offline discussion (in April as far as I remember?) you or @martint specifically requested. If I remember correctly @martint explicitly referenced `NetworkTopology` :) Or do you have something else in mind?

File: presto-main/src/main/java/com/facebook/presto/spiller/SpillerFactoryWithStats.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+
+public abstract class SpillerFactoryWithStats
+        implements SpillerFactory
+{
+    protected final AtomicLong spilledBytes = new AtomicLong();
+
+    @Override
+    @Managed
+    public long getSpilledBytes()
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/SpillerFactoryWithStats.java]: We thought about it and probably at some point we will add this (we have even ticket for that). But baby steps first :-)

**Comment by pnowojski** [comment]
@cberner Please check my responses to your comments.

Regarding limiting spill space. Yes, we are thinking about how to add this (it is probably the most significant missing feature), but (again) I would like to have this as a separate feature/PR. Lets try to keep size of PRs reasonably small (this one already have over 2000 lines) so we don't have maintain and rebase huge chunk of codes for long time.

File: presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java (hunk 4 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java]: Sounds good. Can you file an issue(s) to keep track of the various refactors that need to be done?

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java (hunk 12 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: Yep, `NetworkTopology` is how we used to make things extensible, but now we're going with the approach using for security and resource groups, as it allows extensibility via normal plugins, rather than re-building the server

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSpiller.java (hunk 11 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSpiller.java]: I was thinking we would integrate it with the `OperatorStats`, but we can put that off 'til later

**Comment by cberner** [comment]
I merged these:
* Fix closing HashAggregationBuilder
* Add Spiller and SpillerFactory interface
* Implement BinaryFileSpiller

still looking at the others

File: presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java (hunk 4 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/spiller/Spiller.java]: I have added some issues: https://github.com/prestodb/presto/issues/6258 https://github.com/prestodb/presto/issues/6259 and https://github.com/prestodb/presto/issues/6260. For this particular comment, I would postpone filling issue until spilling for join will go in (or we might refactor Spillers during joins spilling PR). Now it's weird to create an issue about code that even doesn't have PR yet :-)

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java (hunk 12 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: Do you want it to be made as part of this PR or the follow up work?

File: presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java [hunk 14]
```diff
@@ -64,4 +64,6 @@ static GroupByHash createGroupByHash(
     boolean contains(int position, Page page, int[] hashChannels);
 
     int putIfAbsent(int position, Page page);
+
+    int compare(int leftGroupId, int rightGroupId);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java]: I don't think this API makes sense. `GroupByHash` is a hash table, so there's no guarantee that groups are comparable. I realize there will be a performance penalty, but I think for now it would be better to write the data into a `Page` using `appendValuesTo` and then sort the `Page`

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java (hunk 12 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: Either is fine with me. I just don't want to merge this commit the way it is

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    {
+        checkState(spiller.isPresent());
+
+        hashAggregationBuilder.setOutputPartial();
+
+        Iterator<Page> mergedSpilledPages = MergeSort.merge(
+                groupByTypes,
+                hashAggregationBuilder.buildIntermediateTypes(),
+                ImmutableList.<Iterator<Page>>builder()
+                        .addAll(spiller.get().getSpills())
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: calling `getSpills()` will potentially allocate a lot of memory, since it has to allocate a `BufferedInputStream` for every spill. We're going to need to track that and limit it to avoid OOM'ing the workers. Is that something you're planning for a future PR, and if so can you file a issue and add a TODO in the code?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 16]
```diff
@@ -55,6 +56,7 @@
     private int re2JDfaStatesLimit = Integer.MAX_VALUE;
     private int re2JDfaRetries = 5;
     private RegexLibrary regexLibrary = JONI;
+    private DataSize operatorMemoryLimitBeforeSpill = new DataSize(0, DataSize.Unit.MEGABYTE);
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: done

File: presto-main/src/test/java/com/facebook/presto/operator/TestMergeSort.java [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+import static com.facebook.presto.operator.PageAssertions.assertPageEquals;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class TestMergeSort
+{
+    @Test
+    public void testMerge() throws Exception
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/TestMergeSort.java]: fixed

File: presto-main/src/test/java/com/facebook/presto/operator/spiller/BenchmarkBinaryFileSpiller.java [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+        private int pagesCount = 10;
+
+        private List<Page> pages;
+        private Spiller spiller;
+
+        @Setup
+        public void setup()
+                throws ExecutionException, InterruptedException
+        {
+            executor = listeningDecorator(newSingleThreadScheduledExecutor());
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/spiller/BenchmarkBinaryFileSpiller.java]: I neither changed benchmark results nor improved stability but I have replaced it. It might be easier to debug. On the other hand it makes this benchmark a little bit more artificial.

File: presto-main/src/test/java/com/facebook/presto/operator/spiller/TestSpillerFactory.java [hunk 19]
```diff
@@ -52,7 +53,7 @@ public void testOverrideSpillPath()
     {
         File tempDir = Files.createTempDir();
         String spillPath = new File(tempDir, "custom_path").getAbsolutePath();
-        ImmutableMap<String, String> serverProperties = ImmutableMap.of("experimental.spiller-spill-path", spillPath);
+        Map<String, String> serverProperties = ImmutableMap.of("experimental.spiller-spill-path", spillPath);
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/spiller/TestSpillerFactory.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 15 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: I had some greater change in mind in this area later that would involve manual handling of those buffers (and accounting them). But nevertheless I have added issue for this accounting: https://github.com/prestodb/presto/issues/6272

File: presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java (hunk 14 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java]: I can not sort single pages. All positions in all pages produced from `GroupByHash` have to be sorted. So in your approach, I would have to materialize ALL result pages, sort them in memory and then spill them to disk. Problem would be with materialization, which would basically double the memory usage of the operator.

Solutions for that are ugly at best and far from optimal in terms of performance. I don't want to focus on performance with this first iteration, but if there is an easy way to implement something fast, it would be shame to de-optimize code, because later with spilling to highly compressed files stored on SSD should be pretty fast.

Is this `compare` really that  bad?

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java (hunk 12 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: Ok, I have dropped this commit

**Comment by pnowojski** [comment]
Rebased and applied requested changes (beside `GroupByHash::compare` method - please look at my response).

File: presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java (hunk 14 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java]: You don't need to double memory usage. You could spill some constant number of pages at a time.

`compare` seems very out of place since it will fail for types that are not orderable, whereas all the other operations work. Will discuss with a couple other people to get their opinions.

Another option might be to add an API that allows a lambda to be applied to a group id.

**Comment by cberner** [comment]
Merged these:
* Add operator_memory_limit_before_spill property
* Add spill_enabled property
* Add `experimental.spiller-spill-path` property
* Add `experimental.spiller-threads` property
* Add benchmark for BinaryFileSpiller

File: presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java (hunk 14 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java]: Thought about this a bit more. Can we instead do all the sorting based on the hash code? Not only would that fix this issue (we can could `getHashCode` instead of `compare`), but it would also make the spilling code more generic, so that it worked with any type, rather than just ones that are orderable

**Comment by cberner** [comment]
Back to you for question about the `compare` method.

File: presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java (hunk 14 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/GroupByHash.java]: That's a good idea. Changed to `compare` to `getRawHash`.

There are only two drawbacks. First is that result of such `MergeHashSort` is not ordered strictly speaking because of hash collisions. Rows with same hash value can interleaved. It makes it a little be hacky from this perspective, but doesn't change correctness (as long as positions with same hash are merged into one single `Page` - which is happening).

Second drawback is that it comparing is faster then calculating hash for multiple columns - for hash we always need to read/touch all values, but that's shouldn't be that big a deal.

**Comment by pnowojski** [comment]
Back to you :) I have chosen name for `compare` replacement to be `getRawHash` not `getHashCode` as you proposed, because otherwise it might be unclear what hash are we talking about (rawHash from type? murmur(rawHash)? xxhash64(rawHash)? other custom hash?).

File: presto-main/src/main/java/com/facebook/presto/operator/MergeHashSort.java [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+            else {
+                current = new PagePosition(current.getPage(), current.getPosition() + 1);
+            }
+            return current;
+        }
+    }
+
+    /**
+     * This class rewrites iterator over PagePosition to iterator over Pages.
+     * Positions with same hash value are guaranteed to be in the same result page.
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeHashSort.java]: This doesn't sound like a good idea. `Page`s should always be targeted to the maximum size, so that we don't run out of heap space. Why do we need to make this gaurantee?

Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeHashSort.java]: We need it to simplify merging and building final result. This let us assume while reading spills, that every `Page` we get from `MergeHashSort` contains all relevant entries to produce final output. Otherwise, we would have to manually validate whether next `Page` contains positions new keys, or keys that we are already aggregated. Further more, those keys could change in the middle of the `Page`. We could end up with situation that we don't have memory for adding new keys to the `HashAggregationBuilder` but next Page would contain both old keys that we need to release memory and new keys (for which to aggregate we don't have memory).

Secondly, this is shouldn't be big issue:
1. those Pages are immediately consumed and aggregated
2. there shouldn't be that many keys with the same value, because the more keys with same value the less memory is actually needed to perform aggregation, so if we are already spilling that means that we have lots of unique aggregation keys. I would expect this code to only marginally exceed maximum size.
3. Those are already partially aggregated Pages

I would postpone fixing this until we see that this could be an issue, or at least as a separate follow up improvement.

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeHashSort.java]: I disagree, I think this could easily occur in real-world use. For example, suppose you're performing a MAX aggregation on varchars, or any other aggregation that can have large state objects. If you don't have enough memory and have to spill, there could easily end up being a very large number of pages that then have to be merged together.

If it's easiest to keep things in single pages though, we can do that in theory, but it means that this PageBuilder and its Pages need memory tracking to avoid running the VM out of memory. That seems like a more complicated change to me, and also error prone as we generally don't track single Page objects, as they're assumed to be small

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeHashSort.java]: We can schedule a time to discuss the design of this, if that's easiest for you. My mornings are all blocked this week, but I could do next week

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeHashSort.java]: Ok, actually I discussed with @martint and we'll go ahead and merge this as-is. Could you file a follow up issue to fix this though?

**Comment by cberner** [comment]
These look good:
* Add tests for queries using binary spiller
* Add getRawHash method to GroupByHash
* Add MergeHashSort class for merging iterators of pages

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java [hunk 21]
```diff
@@ -88,19 +115,12 @@ public void processPage(Page page)
         for (Aggregator aggregator : aggregators) {
             aggregator.processPage(groupIds, page);
         }
-        updateMemory();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: Why was this removed?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java [hunk 22]
```diff
@@ -148,27 +216,51 @@ protected Page computeNext()
...
+    }
+
+    private Iterator<Integer> consecutiveGroupIds()
+    {
+        return IntStream.range(0, groupByHash.getGroupCount()).iterator();
+    }
+
+    private Iterator<Integer> hashSortedGroupIds()
+    {
+        List<Integer> groupIds = Lists.newArrayList(consecutiveGroupIds());
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: This could materialize GBs worth of memory, and definitely cannot be merged as-is

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: The amount of system memory is configurable, so the user could easily set it to GBs.

I would fix this in one of two ways:
1) limit the number of groups spilled to a small constant
2) account for this array when accepting input, so that there is spare memory left to make this allocation.

Trying to drop memory from GroupByHash seems too complicated

**Comment by pnowojski** [comment]
@cberner I have added memory reservation for sorting group ids. Please check last commit.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java [hunk 23]
```diff
@@ -153,6 +153,15 @@ public long getSizeInMemory()
         return sizeInMemory;
     }
 
+    /**
+     * building hash sorted results requires memory for sorting group IDs.
+     * This method returns size of that memory requirement.
+     */
+    public long getGroupIdsSortingSize()
+    {
+        return getGroupCount() * Integer.BYTES;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: I don't think this is accurate. A `ArrayList<Integer>` uses more than 4 bytes per entry, because elements are `Object`s. Instead I would make this more explicit by pre-allocating the array used by `hashSortedGroupIds` and use `IntBigArray` then you can just use `sizeOf`

Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: Ops, indeed. Sorry, old habits from C++ world. I'm really surprised how tricky it is to actually sort arrays of int in Java... I can not use int[] because of GC AND because there is no `Arrays.sort(int[]) `with a comparator. `IntArrayList` for sorting with comparator also creates temporary `Integer[]`. Same as `IntStream.sorted()`. `IntBigArray` on the other hand has no sort method at all.

I have implemented `List<Integer>` wrapper around `IntBigArray`.

I don't want to pre allocate this array, because it would be unused with disabled spilling.

**Comment by pnowojski** [comment]
Rebased and resolved conflicts with @dain's changes that went in overnight.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+import com.facebook.presto.testing.LocalQueryRunner;
+import com.facebook.presto.tpch.TpchConnectorFactory;
+import com.google.common.collect.ImmutableMap;
+
+import static com.facebook.presto.testing.TestingSession.TESTING_CATALOG;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static com.facebook.presto.tpch.TpchMetadata.TINY_SCHEMA_NAME;
+
+public class TestLocalBinarySpilledQueries
+        extends AbstractTestQueries
```
Comment (arhimondr) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java]: Do we really need to run the entire test suite? Can we run only small subset of the tests that matter? It takes ~10 minutes to run all the tests.

Comment (maciejgrzybek) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java]: You can write a query within existing `AbstractTestQueries` just changing session property for that one particular query.
This overhead with adding yet another `AbstractTestQueries` is unnecessary and adds burden to our automation.

**Comment by pnowojski** [comment]
Added one more commit. Missing `isBlocked` method was causing drivers to block on wrong futures (from different operators) and they were not woken up when spilling has finished.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java (hunk 24 above)
Comment (pnowojski) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java]: It's hard to pick some queries subset to test (for example because it's important to test for different types in columns) and since this feature is pretty intrusive it's worth testing for all queries. During implementation there were bugs that were discovered by some random unexpected query.

Comment (arhimondr) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java]: Then maybe we need to exclude this suite from the tests run for each commit, and run those tests only in presto checks?

**Comment by cberner** [comment]
Merged these two:
* Add getRawHash method to GroupByHash
* Add MergeHashSort class for merging iterators of pages

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java [hunk 25]
```diff
@@ -152,27 +233,82 @@ protected Page computeNext()
...
+        }
+        return types;
+    }
+
+    private Iterator<Integer> consecutiveGroupIds()
+    {
+        return IntStream.range(0, groupByHash.getGroupCount()).iterator();
+    }
+
+    private Iterator<Integer> hashSortedGroupIds()
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: I think we should avoid all the boxing this is going to add. `fastutil` has a `IntIterator` class, as well as sorting methods that operate on primitives

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java [hunk 26]
```diff
@@ -152,27 +233,82 @@ protected Page computeNext()
...
+    public List<Type> buildTypes()
+    {
+        ArrayList<Type> types = new ArrayList<>(groupByHash.getTypes());
+        for (Aggregator aggregator : aggregators) {
+            types.add(aggregator.getType());
+        }
+        return types;
+    }
+
+    private Iterator<Integer> consecutiveGroupIds()
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: Same comment as below

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/MergingHashAggregationBuilder.java [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+            }
+
+            @Override
+            public Page next()
+            {
+                if (!resultPages.hasNext()) {
+                    rebuildHashAggregationBuilder();
+                    long memorySize = 0; // ensure that at least one merged page will be processed
+
+                    while (sortedPages.hasNext() && !shouldProduceOutput(memorySize)) {
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/MergingHashAggregationBuilder.java]: There was such comment over `PageRewriteIterator` in `MergeHashSort`, but I have moved to `merge` method.

**Comment by pnowojski** [comment]
Done

**Comment by pnowojski** [comment]
Thanks :)
